Use NNPACK for strided convolutions.

AshkanAliabadi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#27402 Use NNPACK for strided convolutions.**

Differential Revision: [D18126265](https://our.internmc.facebook.com/intern/diff/D18126265)

Commit:
Use NNPACK for strided convolutions.

[ghstack-poisoned]

Commit:
Update on "Use NNPACK for strided convolutions."




[ghstack-poisoned]

Commit:
Update on "Use NNPACK for strided convolutions."




[ghstack-poisoned]

Commit:
Update on "Use NNPACK for strided convolutions."




[ghstack-poisoned]

dzhulgakov commented on aten/src/ATen/native/Convolution.cpp:
```diff
@@ -648,7 +647,7 @@ at::Tensor _convolution(
     if (params.use_cpu_depthwise3x3_winograd(input, weight)) {
       output = convolution_depthwise3x3_winograd_stub(
-        input.device().type(), input, weight, bias, params.padding, params.stride, params.groups);
+        input.device().type(), input, weight, bias, params.stride, params.padding, params.groups);
```
wait, was it a bug? do we have a unittest for it?

AshkanAliabadi replied:
This is a crazy one.  Here's my explanation of how I think it was working: https://github.com/pytorch/pytorch/pull/27117#discussion_r331209266

Commit:
Update on "Use NNPACK for strided convolutions."




[ghstack-poisoned]

Commit:
Update on "Use NNPACK for strided convolutions."




[ghstack-poisoned]

AshkanAliabadi commented on aten/src/ATen/CMakeLists.txt:
```diff
@@ -211,7 +211,7 @@ endif()
 if(AT_NNPACK_ENABLED)
   include_directories(${NNPACK_INCLUDE_DIRS})
-  list(APPEND ATen_CPU_DEPENDENCY_LIBS nnpack) # cpuinfo is added below
+  list(APPEND ATen_CPU_DEPENDENCY_LIBS nnpack nnpack_reference_layers) # cpuinfo is added below
```
NNPACK does not support strided convolutions on batches (used during training,) hence we fallback to the slower reference implementation here.

AshkanAliabadi commented on aten/src/ATen/native/NNPACK.cpp:
```diff
@@ -243,16 +252,35 @@ Tensor _nnpack_spatial_convolution(
+          weight.data_ptr<float>(),
+          bias_.data_ptr<float>(),
+          output.data_ptr<float>(),
+          nnpack_threadpool());
```
NNPACK does not support strided convolutions on batches (used during training,) hence we fallback to the slower reference implementation here.

AshkanAliabadi reviewed (commented):
Pending CI, it seems that I finally managed to get this working.  This is really not an ideal solution in my opinion and definitely not the direction we should be taking PyTorch mobile in, in the long run.  I think the ideal solution would be to have one single mobile-focused backend, just like the reference or vendor-provided backends we currently have, that handles floating point and quantized operations both for NHWC and NCHW layout but efficiently for mobile.  That is a considerable undertaking though, and if we want to close the gap in the meanwhile, we need short-term solutions like this unfortunately.

Commit:
Update on "Use NNPACK for strided convolutions."




[ghstack-poisoned]

Commit:
Update on "Use NNPACK for strided convolutions."


Differential Revision: [D18126265](https://our.internmc.facebook.com/intern/diff/D18126265)

[ghstack-poisoned]

Commit:
Update on "Use NNPACK for strided convolutions."


Differential Revision: [D18126265](https://our.internmc.facebook.com/intern/diff/D18126265)

[ghstack-poisoned]

Commit:
Update on "Use NNPACK for strided convolutions."


Differential Revision: [D18126265](https://our.internmc.facebook.com/intern/diff/D18126265)

[ghstack-poisoned]

Commit:
Update on "Use NNPACK for strided convolutions."


Differential Revision: [D18126265](https://our.internmc.facebook.com/intern/diff/D18126265)

[ghstack-poisoned]

Commit:
Update on "Use NNPACK for strided convolutions."


Differential Revision: [D18126265](https://our.internmc.facebook.com/intern/diff/D18126265)

[ghstack-poisoned]

Commit:
Update on "Use NNPACK for strided convolutions."


Differential Revision: [D18126265](https://our.internmc.facebook.com/intern/diff/D18126265)

[ghstack-poisoned]

AshkanAliabadi commented:
Dependent on https://github.com/Maratyszcza/NNPACK/pull/172.

Commit:
Update on "Use NNPACK for strided convolutions."


Differential Revision: [D18126265](https://our.internmc.facebook.com/intern/diff/D18126265)

[ghstack-poisoned]

Commit:
Update on "Use NNPACK for strided convolutions."


Differential Revision: [D18126265](https://our.internmc.facebook.com/intern/diff/D18126265)

[ghstack-poisoned]

Commit:
Update on "Use NNPACK for strided convolutions."


Differential Revision: [D18126265](https://our.internmc.facebook.com/intern/diff/D18126265)

[ghstack-poisoned]
